File: apps/trackr-mobile-ios/trackr-mobile-ios/Models/Taxonomy.swift
//
//  Taxonomy.swift
//  trackr-mobile-ios
//
//  Mirror of web/src/lib/config/taxonomy.ts — labels and colors must stay
//  in sync with the web app.
//

import SwiftUI

enum TaskStatus: String, CaseIterable, Identifiable {
    case backlog, todo, inProgress, paused, inReview, done

    var id: String { rawValue }

    var label: String {
        switch self {
        case .backlog: "Backlog"
        case .todo: "Todo"
        case .inProgress: "In Progress"
        case .paused: "Paused"
        case .inReview: "In Review"
        case .done: "Done"
        }
    }

    var color: Color {
        switch self {
        case .backlog: Color(hex: 0x7C7C84)
        case .todo: Color(hex: 0x9AA4B2)
        case .inProgress: Color(hex: 0xF0A85C)
        case .paused: Color(hex: 0xE9C46A)
        case .inReview: Color(hex: 0xB591E3)
        case .done: Color(hex: 0x7FC8A9)
        }
    }
}

enum TaskPriority: String, CaseIterable, Identifiable {
    case none, low, medium, high, urgent

    var id: String { rawValue }

    var label: String {
        switch self {
        case .none: "None"
        case .low: "Low"
        case .medium: "Medium"
        case .high: "High"
        case .urgent: "Urgent"
        }
    }

    /// Number of lit bars in the priority indicator (0–4, web parity).
    var bars: Int {
        switch self {
        case .none: 0
        case .low: 1
        case .medium: 2
        case .high: 3
        case .urgent: 4
        }
    }

    var color: Color {
        switch self {
        case .none: Color(hex: 0x5B5B62)
        case .low: Color(hex: 0x7A9CF0)
        case .medium: Color(hex: 0xF0A85C)
        case .high: Color(hex: 0xEF7A6D)
        case .urgent: Color(hex: 0xEF4F5E)
        }
    }
}

enum TaskType: String, CaseIterable, Identifiable {
    case task, bug, improvement, feature, chore

    var id: String { rawValue }

    var label: String {
        switch self {
        case .task: "Task"
        case .bug: "Bug"
        case .improvement: "Improvement"
        case .feature: "Feature"
        case .chore: "Chore"
        }
    }

    var color: Color {
        switch self {
        case .task: Color(hex: 0x7A9CF0)
        case .bug: Color(hex: 0xEF7A6D)
        case .improvement: Color(hex: 0xEF7A6D)
        case .feature: Color(hex: 0x7FC8A9)
        case .chore: Color(hex: 0xC08BD6)
        }
    }

    var systemImage: String {
        switch self {
        case .task: "square"
        case .bug: "ladybug"
        case .improvement: "triangle"
        case .feature: "sparkle"
        case .chore: "gearshape"
        }
    }
}

enum DueTone {
    case overdue, urgent, soon, normal

    var color: Color? {
        switch self {
        case .overdue: Color(hex: 0xEF4F5E)
        case .urgent: .accentColor
        case .soon: Color(hex: 0xD8A24A)
        case .normal: nil
        }
    }
}

/// Mirror of web TICKET_STATUSES in taxonomy.ts.
enum TicketStatus: String, CaseIterable, Identifiable {
    case open, inProgress, waitingOnCustomer, waitingOnAgent, paused, resolved, closed

    var id: String { rawValue }

    var label: String {
        switch self {
        case .open: "Open"
        case .inProgress: "In Progress"
        case .waitingOnCustomer: "Waiting on customer"
        case .waitingOnAgent: "Waiting on agent"
        case .paused: "Paused"
        case .resolved: "Resolved"
        case .closed: "Closed"
        }
    }

    var color: Color {
        switch self {
        case .open: Color(hex: 0x7A9CF0)
        case .inProgress: Color(hex: 0xF0A85C)
        case .waitingOnCustomer: Color(hex: 0xB591E3)
        case .waitingOnAgent: Color(hex: 0xEF7A6D)
        case .paused: Color(hex: 0xE9C46A)
        case .resolved: Color(hex: 0x7FC8A9)
        case .closed: Color(hex: 0x7C7C84)
        }
    }

    var isClosed: Bool { self == .resolved || self == .closed }
}

/// Mirror of web TICKET_CATEGORIES in taxonomy.ts.
enum TicketCategory: String, CaseIterable, Identifiable {
    case general, billing, technicalIssue, featureRequest

    var id: String { rawValue }

    var label: String {
        switch self {
        case .general: "General"
        case .billing: "Billing"
        case .technicalIssue: "Technical issue"
        case .featureRequest: "Feature request"
        }
    }

    var color: Color {
        switch self {
        case .general: Color(hex: 0x9AA4B2)
        case .billing: Color(hex: 0xE9C46A)
        case .technicalIssue: Color(hex: 0xEF7A6D)
        case .featureRequest: Color(hex: 0x7FC8A9)
        }
    }
}

/// Mirror of web TICKET_CHANNELS in taxonomy.ts.
enum TicketChannel: String, CaseIterable, Identifiable {
    case webForm, email, chat, api

    var id: String { rawValue }

    var label: String {
        switch self {
        case .webForm: "Web form"
        case .email: "Email"
        case .chat: "Chat"
        case .api: "API"
        }
    }

    var systemImage: String {
        switch self {
        case .webForm: "globe"
        case .email: "envelope"
        case .chat: "bubble.left.and.bubble.right"
        case .api: "terminal"
        }
    }
}

extension TaskPriority {
    /// Tickets reuse the trackr priorities minus 'none' — every ticket has
    /// a priority (web parity: TICKET_PRIORITIES).
    static var ticketCases: [TaskPriority] {
        allCases.filter { $0 != .none }
    }
}

/// Estimate presets (minutes) shared by the create sheet and detail view.
enum EstimateOptions {
    static let all: [(minutes: Int?, label: String)] = [
        (nil, "None"), (15, "15m"), (30, "30m"), (60, "1h"),
        (120, "2h"), (180, "3h"), (240, "4h"), (360, "6h"), (480, "1d"),
    ]
}

/// Mirror of web PROJECT_STATUS in taxonomy.ts.
enum ProjectStatus: String, CaseIterable, Identifiable {
    case prospect, planned, active, paused, completed, cancelled, archived

    var id: String { rawValue }

    var label: String {
        switch self {
        case .prospect: "Prospect"
        case .planned: "Planned"
        case .active: "Active"
        case .paused: "Paused"
        case .completed: "Completed"
        case .cancelled: "Cancelled"
        case .archived: "Archived"
        }
    }

    var color: Color {
        switch self {
        case .prospect: Color(hex: 0xC08BD6)
        case .planned: Color(hex: 0x7A9CF0)
        case .active: Color(hex: 0x7FC8A9)
        case .paused: Color(hex: 0x9AA4B2)
        case .completed: Color(hex: 0x7C9B86)
        case .cancelled: Color(hex: 0xEF7A6D)
        case .archived: Color(hex: 0x9AA4B2)
        }
    }
}

File: web/src/lib/server/mcp/ui/shared/taxonomy.ts
// Display metadata for statuses, priorities, types and categories — a copy of
// what src/lib/config/taxonomy.ts declares, kept dependency-free because this
// code is bundled into the widget iframe, not the app.

export type StatusMeta = { label: string; dot: string; tint: string };

export const TICKET_STATUS: Record<string, StatusMeta> = {
	open: { label: 'Open', dot: '#7a9cf0', tint: 'rgba(122,156,240,0.14)' },
	in_progress: { label: 'In progress', dot: '#f0a85c', tint: 'rgba(240,168,92,0.16)' },
	waiting_on_customer: {
		label: 'Waiting on customer',
		dot: '#b591e3',
		tint: 'rgba(181,145,227,0.14)'
	},
	waiting_on_agent: { label: 'Waiting on agent', dot: '#ef7a6d', tint: 'rgba(239,122,109,0.14)' },
	paused: { label: 'Paused', dot: '#e9c46a', tint: 'rgba(233,196,106,0.14)' },
	resolved: { label: 'Resolved', dot: '#7fc8a9', tint: 'rgba(127,200,169,0.16)' },
	closed: { label: 'Closed', dot: '#7c7c84', tint: 'rgba(124,124,132,0.14)' }
};

export const TASK_STATUS: Record<string, StatusMeta> = {
	backlog: { label: 'Backlog', dot: '#7c7c84', tint: 'rgba(124,124,132,0.14)' },
	todo: { label: 'Todo', dot: '#9aa4b2', tint: 'rgba(154,164,178,0.14)' },
	in_progress: { label: 'In progress', dot: '#f0a85c', tint: 'rgba(240,168,92,0.16)' },
	paused: { label: 'Paused', dot: '#e9c46a', tint: 'rgba(233,196,106,0.14)' },
	in_review: { label: 'In review', dot: '#b591e3', tint: 'rgba(181,145,227,0.16)' },
	done: { label: 'Done', dot: '#7fc8a9', tint: 'rgba(127,200,169,0.16)' }
};

export const PROJECT_STATUS: Record<string, StatusMeta> = {
	prospect: { label: 'Prospect', dot: '#9aa4b2', tint: 'rgba(154,164,178,0.14)' },
	planned: { label: 'Planned', dot: '#7a9cf0', tint: 'rgba(122,156,240,0.14)' },
	active: { label: 'Active', dot: '#7fc8a9', tint: 'rgba(127,200,169,0.16)' },
	paused: { label: 'Paused', dot: '#e9c46a', tint: 'rgba(233,196,106,0.14)' },
	completed: { label: 'Completed', dot: '#b591e3', tint: 'rgba(181,145,227,0.16)' },
	cancelled: { label: 'Cancelled', dot: '#ef7a6d', tint: 'rgba(239,122,109,0.14)' },
	archived: { label: 'Archived', dot: '#7c7c84', tint: 'rgba(124,124,132,0.14)' }
};

export const SEARCH_TYPE: Record<string, StatusMeta> = {
	ticket: { label: 'Ticket', dot: '#ef7a6d', tint: 'rgba(239,122,109,0.14)' },
	task: { label: 'Task', dot: '#7a9cf0', tint: 'rgba(122,156,240,0.14)' },
	project: { label: 'Project', dot: '#7fc8a9', tint: 'rgba(127,200,169,0.16)' },
	wiki: { label: 'Wiki', dot: '#c08bd6', tint: 'rgba(192,139,214,0.16)' },
	note: { label: 'Note', dot: '#e9c46a', tint: 'rgba(233,196,106,0.14)' }
};

export const PRIORITY: Record<string, { label: string; bars: number; color: string }> = {
	none: { label: 'None', bars: 0, color: '#5b5b62' },
	low: { label: 'Low', bars: 1, color: '#7a9cf0' },
	medium: { label: 'Medium', bars: 2, color: '#f0a85c' },
	high: { label: 'High', bars: 3, color: '#ef7a6d' },
	urgent: { label: 'Urgent', bars: 4, color: '#ef4f5e' }
};

export const TASK_TYPE: Record<string, { label: string; color: string }> = {
	task: { label: 'Task', color: '#7a9cf0' },
	bug: { label: 'Bug', color: '#ef7a6d' },
	improvement: { label: 'Improvement', color: '#ef7a6d' },
	feature: { label: 'Feature', color: '#7fc8a9' },
	chore: { label: 'Chore', color: '#c08bd6' }
};

export const TICKET_CATEGORY: Record<string, { label: string; color: string }> = {
	general: { label: 'General', color: '#9aa4b2' },
	billing: { label: 'Billing', color: '#e9c46a' },
	technical_issue: { label: 'Technical issue', color: '#ef7a6d' },
	feature_request: { label: 'Feature request', color: '#7fc8a9' }
};

export const FALLBACK_STATUS: StatusMeta = {
	label: '',
	dot: '#7c7c84',
	tint: 'rgba(124,124,132,0.14)'
};

export function statusMeta(map: Record<string, StatusMeta>, id: string): StatusMeta {
	return map[id] ?? { ...FALLBACK_STATUS, label: id.replace(/_/g, ' ') };
}
